src/magicbox/QueriesProcessor.ts

```diff
@@ -0,0 +1,91 @@
+import { defaults } from 'underscore';
+
+export interface IQueriesProcessorOptions {
+  timeout: number;
+}
+
+export enum QueryProcessResultStatus {
+  Finished = 'finished',
+  TimedOut = 'timedout',
+  Overriden = 'overriden'
+}
+
+export type QueryProcessResult<ItemType> =
+  | {
+      status: QueryProcessResultStatus.Finished | QueryProcessResultStatus.TimedOut;
+      items: ItemType[];
+    }
+  | {
+      status: QueryProcessResultStatus.Overriden;
+    };
+
+export class QueriesProcessor<ItemType> {
+  private static async streamQueries<ItemType>(queries: Promise<ItemType[]>[], output: ItemType[]) {
+    await Promise.all(queries.map(query => query.then(items => output.push(...items))));
+  }
+
+  private override: () => any;
+  private options: IQueriesProcessorOptions;
+
+  constructor(options: Partial<IQueriesProcessorOptions> = {}) {
+    this.options = defaults(options, <IQueriesProcessorOptions>{
+      timeout: 500
+    });
+  }
+
+  /**
+   * Overrides the previous queries and accumulates the result of promise arrays with a timeout.
+   */
+  public async processQueries(queries: (ItemType[] | Promise<ItemType[]>)[]): Promise<QueryProcessResult<ItemType>> {
+    this.overrideIfProcessing();
+    let items: ItemType[] = [];
+
+    const asynchronousQueries: Promise<ItemType[]>[] = [];
+    for (const query of queries) {
+      if (query instanceof Promise) {
+        asynchronousQueries.push(query);
+      } else {
+        items.push(...query);
+      }
+    }
+
+    let status: QueryProcessResultStatus;
+    if (asynchronousQueries.length > 0) {
+      status = await Promise.race([
+        QueriesProcessor.streamQueries(asynchronousQueries, items).then(() => QueryProcessResultStatus.Finished),
+        this.waitForOverride().then(() => QueryProcessResultStatus.Overriden),
```

Comment:
I'm curious why we have the override and timeout options here? Wouldn't we want to succeed as much as possible?

Reply:
Since this is going to be used to replace `SuggestionsManager.mergeSuggestions`, the `override` option is meant as an equivalent behaviour of `SuggestionsManager`'s `reject('new request queued')`.